### PR TITLE
feat(clp-s::timestamp_parser): Add support for marshalling timestamps.

### PR DESCRIPTION
**Pull request by gibber9809**
# Description

This PR follows up on #1626, #1599, #1564, and #1385 by implementing marshalling for numeric and date-time timestamp patterns. The marshalling implementation currently uses `fmt` to format some parts of a timestamp, which results in some extra copying of small strings; if this ends up being a performance problem we can switch to a custom implementation that elides these small copies.

This PR may be followed up by one more PR to the `clp-s::timestamp_parser` library to improve support for parsing leap-seconds before a PR to fully integrate `clp-s::timestamp_parser` with clp-s.

# Checklist

* [x] The PR satisfies the [contribution guidelines][yscope-contrib-guidelines].
* [x] This is a breaking change and that has been indicated in the PR title, OR this isn't a
  breaking change.
* [x] Necessary docs have been updated, OR no docs need to be updated.

# Validation performed

* Added assertions to verify that all timestamps that we parse in the tests can be marshalled correctly.

[yscope-contrib-guidelines]: https://docs.yscope.com/dev-guide/contrib-guides-overview.html

## Summary by CodeRabbit

* **New Features**
  * Added timestamp marshaling/formatting to produce date‑time and numeric epoch outputs (12/24h, AM/PM, day/month/year, day‑of‑week, timezone, subsecond precision).

* **Bug Fixes**
  * Improved validation for zero‑padding and clearer errors for invalid/incompatible timestamp patterns.

* **Tests**
  * Added unit tests that parse timestamps and verify round‑trip formatting for several patterns and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: components/core/src/clp_s/timestamp_parser/TimestampParser.cpp [hunk 1]
```diff
@@ -243,6 +248,38 @@ find_first_matching_prefix(std::string_view str, std::span<std::string_view cons
...
+ * @param pattern
+ * @param buffer The buffer that the marshalled timestamp is appended to.
+ * @return A void result on success, or an error code indicating the failure:
+ * - ErrorCodeEnum::InvalidTimestampPattern if `pattern` contains format malformed format
+ *   specifiers, or format specifiers that aren't supported in date-time timestamps.
+ * - ErrorCodeEnum::IncompatibleTimestampPattern if `timestamp` can not be represented by `pattern`.
+ */
+[[nodiscard]] auto marshal_date_time_timestamp(
+        epochtime_t timestamp,
+        TimestampPattern const& timestamp_pattern,
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/timestamp_parser/TimestampParser.cpp]: `pattern` to match the docstring? lol
(I prefer `pattern` for clarity

File: components/core/src/clp_s/timestamp_parser/TimestampParser.cpp [hunk 2]
```diff
@@ -243,6 +248,38 @@ find_first_matching_prefix(std::string_view str, std::span<std::string_view cons
...
+ * @param pattern
+ * @param buffer The buffer that the marshalled timestamp is appended to.
+ * @return A void result on success, or an error code indicating the failure:
+ * - ErrorCodeEnum::InvalidTimestampPattern if `pattern` contains format specifiers that aren't
+ *   supported in numeric timestamps.
+ * - ErrorCodeEnum::IncompatibleTimestampPattern if `timestamp` can not be represented by `pattern`.
+ */
+[[nodiscard]] auto marshal_numeric_timestamp(
+        epochtime_t timestamp,
+        TimestampPattern const& timestamp_pattern,
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/timestamp_parser/TimestampParser.cpp]: Same (also for the implementation)

File: components/core/src/clp_s/timestamp_parser/TimestampParser.cpp [hunk 3]
```diff
@@ -445,6 +482,306 @@ auto estimate_timestamp_precision(int64_t timestamp) -> std::pair<int64_t, char>
     };
     return std::make_pair(cFactor, 'E');
 }
+
+auto extract_absolute_subsecond_nanoseconds(epochtime_t timestamp) -> epochtime_t {
+    constexpr auto cFactor{
+            cPowersOfTen
+                    [cNumNanosecondPrecisionSubsecondDigits - cNumSecondPrecisionSubsecondDigits]
+    };
+    auto subsecond_nanoseconds = timestamp % cFactor;
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/timestamp_parser/TimestampParser.cpp]:
```suggestion
    auto const subsecond_nanoseconds{timestamp % cFactor};
```

File: components/core/src/clp_s/timestamp_parser/TimestampParser.cpp [hunk 4]
```diff
@@ -445,6 +482,306 @@ auto estimate_timestamp_precision(int64_t timestamp) -> std::pair<int64_t, char>
...
+            buffer.push_back(c);
+            continue;
+        }
+        escaped = false;
+        switch (c) {
+            case 'E': {  // Epoch seconds.
+                constexpr auto cFactor{cPowersOfTen
+                                               [cNumNanosecondPrecisionSubsecondDigits
+                                                - cNumSecondPrecisionSubsecondDigits]};
+                int64_t epoch_second_timestamp{timestamp / cFactor};
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/timestamp_parser/TimestampParser.cpp]:
```suggestion
                int64_t const epoch_second_timestamp{timestamp / cFactor};
```

File: components/core/src/clp_s/timestamp_parser/TimestampParser.cpp [hunk 5]
```diff
@@ -445,6 +482,306 @@ auto estimate_timestamp_precision(int64_t timestamp) -> std::pair<int64_t, char>
...
+                buffer.append(fmt::format("{}", epoch_second_timestamp));
+                break;
+            }
+            case 'L': {  // Epoch milliseconds.
+                constexpr auto cFactor{cPowersOfTen
+                                               [cNumNanosecondPrecisionSubsecondDigits
+                                                - cNumMillisecondPrecisionSubsecondDigits]};
+                int64_t epoch_millisecond_timestamp{timestamp / cFactor};
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/timestamp_parser/TimestampParser.cpp]:
```suggestion
                int64_t const epoch_millisecond_timestamp{timestamp / cFactor};
```

File: components/core/src/clp_s/timestamp_parser/TimestampParser.cpp [hunk 6]
```diff
@@ -445,6 +482,306 @@ auto estimate_timestamp_precision(int64_t timestamp) -> std::pair<int64_t, char>
...
+                buffer.append(fmt::format("{}", epoch_millisecond_timestamp));
+                break;
+            }
+            case 'C': {  // Epoch microseconds.
+                constexpr auto cFactor{cPowersOfTen
+                                               [cNumNanosecondPrecisionSubsecondDigits
+                                                - cNumMicrosecondPrecisionSubsecondDigits]};
+                int64_t epoch_microsecond_timestamp{timestamp / cFactor};
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/timestamp_parser/TimestampParser.cpp]:
```suggestion
                int64_t const epoch_microsecond_timestamp{timestamp / cFactor};
```

File: components/core/src/clp_s/timestamp_parser/TimestampParser.cpp [hunk 7]
```diff
@@ -229,6 +229,11 @@ find_first_matching_prefix(std::string_view str, std::span<std::string_view cons
 [[nodiscard]] auto extract_timezone_offset_in_minutes(std::string_view str)
         -> ystdlib::error_handling::Result<std::pair<std::string_view, int>>;
 
+/**
+ * Extracts the absolute value of the subsecond fractional component of a timestamp.
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/timestamp_parser/TimestampParser.cpp]:
```suggestion
 * @return The absolute value of the subsecond fractional component of a timestamp.
```

File: components/core/src/clp_s/timestamp_parser/test/test_TimestampParser.cpp [hunk 8]
```diff
@@ -494,7 +496,8 @@ TEST_CASE("timestamp_parser_parse_timestamp", "[clp-s][timestamp-parser]") {
                  1'765'602'000'000'000},
                 {"Jan 21 11:56:42 UTC+0130",
                  R"(\b \d \H:\M:\S UTC\z{+0130})",
-                 1'765'602'000'000'000}
+                 1'765'602'000'000'000},
+                {"1895-11-20T21:55:46,010", R"(\Y-\m-\dT\H:\M:\S,\3)", -2'338'769'053'990'000'000}
```
Comment (LinZhihao-723) [inline components/core/src/clp_s/timestamp_parser/test/test_TimestampParser.cpp:501]: Logs before semiconductor, lol

**Review by LinZhihao-723 — CHANGES_REQUESTED** [review]
The implementation lgtm (also I think the unit test should cover the actual functionality correctness)
Some small style changes
